### PR TITLE
Avoid enforcing `shuffle=False` for eval dataloaders

### LINKED ISSUE #11054: Setting `shuffle=True` for `val_dataloader` doesn't work when `ddp` enabled.
## 🐛 Bug

Setting `shuffle=True` for `val_dataloader` doesn't work when `ddp` enabled.

### To Reproduce

minimum snippet:

```python
from pytorch_lightning import LightningModule, Trainer
from torch import nn
from torch.optim import Adam
from torch.utils.data import DataLoader


class PlModule(LightningModule):
    def __init__(self):
        super().__init__()
        self.model = nn.Linear(1, 1)
        self.train_dataset = list(range(1000))
        self.val_dataset = list(range(1000))

    def configure_optimizers(self):
        return Adam(self.model.parameters(), lr=0.1)

    def training_step(self, batch, *args, **kwargs):
        return self.model(batch)

    def validation_step(self, batch, *args, **kwargs):
        print(f'{batch=}')
        print(f'{self.trainer.val_dataloaders[0].sampler.shuffle=}')
        exit(0)

    def train_dataloader(self):
        return DataLoader(self.train_dataset, shuffle=True, batch_size=5)

    def val_dataloader(self):
        return DataLoader(self.val_dataset, shuffle=True, batch_size=5)



if __name__ == '__main__':
    module = PlModule()
    trainer = Trainer(
        limit_train_batches=0.0,
        limit_val_batches=1.0,
        num_sanity_val_steps=-1,
        gpus=2,
        strategy='ddp',
        enable_model_summary=False
    )
    trainer.fit(module)
```

By execute the snippet above, we get the stdout:

```
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
initializing distributed: GLOBAL_RANK: 1, MEMBER: 2/2
initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/2
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 2 processes
----------------------------------------------------------------------------------------------------

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]
LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]
Validation sanity check: 0it [00:00, ?it/s]/home/balabala/miniconda3/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py:452: UserWarning: Your `val_dataloader` has `shuffle=True`,it is strongly recommended that you turn this off for val/test/predict dataloaders.
  rank_zero_warn(
/home/balabala/miniconda3/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py:110: UserWarning: The dataloader, val_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 96 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Validation sanity check:   0%|                          | 0/100 [00:00<?, ?it/s]batch=tensor([1, 3, 5, 7, 9], device='cuda:1')
self.trainer.val_dataloaders[0].sampler.shuffle=False
batch=tensor([0, 2, 4, 6, 8], device='cuda:0')
self.trainer.val_dataloaders[0].sampler.shuffle=False
```

### Expected behavior

the stdout should contains the follow:

```
batch= # some random batch
self.trainer.val_dataloaders[0].sampler.shuffle=True
```

### Environment

- PyTorch Lightning Version 1.5.2
- PyTorch Version 1.10.0
- Python version 3.8.10

cc @borda @tchaton @rohitgr7 @justusschock @awaelchli @ninginthecloud @kaushikb11 @akihironitta

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #11054

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 1]
```diff
@@ -435,7 +435,13 @@ def _reset_eval_dataloader(
             )
 
         # add samplers
-        dataloaders = [self._prepare_dataloader(dl, False, mode=mode) for dl in dataloaders if dl is not None]
+        dataloaders = [
+            self._prepare_dataloader(
+                dl, shuffle=hasattr(dl, "sampler") and isinstance(dl.sampler, RandomSampler), mode=mode
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:440]: Shouldn't this use the same conditions as `_check_eval_shuffling`?

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/data_connector.py:440]: yes, though so but if its a custom sampler, then it would set shuffle=True, which might be incorrect.

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/data_connector.py:440]: ok, looks like we raise an error in the case of the custom sampler already. will update.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1773,7 +1773,6 @@ def reset_train_dataloader(self, model: Optional["pl.LightningModule"] = None) -
             self.train_dataloader,
             DataLoader,
             self._data_connector._prepare_dataloader,
-            shuffle=True,
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:1776]: reviewer: this is removed from here and is being set inside `_prepare_dataloader`, because we can't check for its value here in case of a `CombinedLoader`.

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
